File: python/0369_Plus_One_Linked_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def plusOne(self, head: ListNode) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        not_nine = dummy
        while head:
            if head.val != 9:
                not_nine = head
            head = head.next
        not_nine.val += 1
        not_nine = not_nine.next
        while not_nine:
            not_nine.val = 0
            not_nine = not_nine.next
        return dummy if dummy.val else dummy.next

# O(N) Time
# O(1) Space

File: cpp/0369_Plus_One_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* plusOne(ListNode* head) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* notNine = dummy;
        while (head) {
            if (head->val != 9) notNine = head;
            head = head->next;
        }
        notNine->val++;
        notNine = notNine->next;
        while (notNine) {
            notNine->val = 0;
            notNine = notNine->next;
        }
        delete notNine;
        return dummy->val ? dummy : dummy->next;
    }
};

// O(N) Time
// O(1) Space